File: leetcode/112. Path Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        
        if root is None:
            return False
        
        def preorder(node, curr):
            if node is None:
                return False
            if node.left is None and node.right is None:
                if curr + node.val == targetSum:
                    return True
                else:
                    return False
            
            return preorder(node.left, curr + node.val) or preorder(node.right, curr + node.val)
            
        
        return preorder(root, 0)

File: leetcode/112. Path Sum.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun hasPathSum(root: TreeNode?, targetSum: Int): Boolean {
        if (root == null) return false
        val remain = targetSum - root.`val`
        if (remain == 0 && root.left == null && root.right == null) return true
        return hasPathSum(root.left, remain) || hasPathSum(root.right, remain)

    }
}
